File: src/contests/cg/puzzle/cpp/teds-compiler.cpp
// 🎮 CodinGame Puzzle - teds-compiler
// https://www.codingame.com/training/easy/teds-compiler

#include <iostream>
#include <string>
using namespace std;

int main() {
    string line;
    getline(cin, line);
    while (!line.empty() && (line.back() == '\r' || line.back() == ' ' || line.back() == '\n'))
        line.pop_back();
    int balance = 0, best = 0;
    for (size_t i = 0; i < line.size(); i++) {
        if (line[i] == '<') balance++;
        else balance--;
        if (balance < 0) break;
        if (balance == 0) best = (int)i + 1;
    }
    cout << best << endl;
}

File: src/contests/cg/puzzle/haskell/teds-compiler.hs
-- 🎮 CodinGame Puzzle - teds-compiler
-- https://www.codingame.com/training/easy/teds-compiler

main :: IO ()
main = do
  line <- getLine
  let go _ _ best [] = best
      go i balance best (c : cs)
        | nb < 0 = best
        | nb == 0 = go (i + 1) nb (i + 1) cs
        | otherwise = go (i + 1) nb best cs
        where
          nb = if c == '<' then balance + 1 else balance - 1
  print (go (0 :: Int) (0 :: Int) (0 :: Int) line)
